tests/AllowHalfOpenTest.php

```diff
@@ -0,0 +1,143 @@
+<?php
+
+namespace React\Tests\Socket;
+
+use React\EventLoop\Factory;
+use React\EventLoop\LoopInterface;
+use React\Socket\ConnectionInterface;
+use React\Socket\Connector;
+use React\Socket\TcpServer;
+
+function _log ($message) {
+    echo $message . "\n";
+};
+
+/**
+ * Class AllowHalfOpenTest
+ * Inspired by https://github.com/nodejs/node/blob/01c680b92aed62e128aa2abd4ad923f9a6a0331a/test/parallel/test-net-connect-options-allowhalfopen.js
+ * Run with `php vendor/bin/phpunit --filter AllowHalfOpenTest`
+ *
+ * @package React\Tests\Socket
+ */
+class AllowHalfOpenTest extends TestCase
+{
+    /**
+     * @var LoopInterface
+     */
+    private $loop;
+    private $server;
+    /**
+     * @var TcpServer
+     */
+    private $socket;
+    private $options = ['allowHalfOpen' => true];
+
+    private $clientReceivedFIN = 0;
+    private $serverConnections = 0;
+    private $clientSentFIN = 0;
+    private $serverReceivedFIN = 0;
+
+    private function createLoop()
+    {
+        return Factory::create();
+    }
+
+    /**
+     * @covers React\Socket\TcpServer::__construct
+     * @covers React\Socket\TcpServer::getAddress
+     */
+    public function setUp()
+    {
+        $this->loop = $this->createLoop();
+        $this->socket = new TcpServer(0, $this->loop, $this->options);
+    }
+
+    /**
+     * @covers React\Socket\TcpServer::handleConnection
+     */
+    public function testAll()
```

Comment:
This is apparently supposed to test the server sending a `FIN` to the client which then sends a `FIN` in return?

I think we should add another test for a client sending a `FIN` to the server which then sends a `FIN` in return.

Also, I don't see this covered in the code, have you verified the `FIN` flags are actually sent?